content/blog/2022-03-22-type-alias/index.md

```diff
@@ -0,0 +1,210 @@
+
+### Open vs Closed World Assumptions
+
+One advantage of performing alias analysis in type-safe languages like Modula-3 and Java is stronger type-safety assumptions about unavailable code. Most user-defined types are encapsulated within a module, and therefore references to those types cannot be aliased by code outside of the module. Figure 12 below shows that as a result, RLE is minimally impacted by open vs closed world assumptions.
+
+<p align="center"> <img src="figure_12.png"/> </p>
+
+The limitations of alias analysis in unsafe languages like C++ with an open world assumption can be mitigated by a proper understanding of undefined behavior.
+
+### Evaluation Summary
+
+TBAA has surprisingly high accuracy in real-world optimizations while maintaining a fast time bound. The evaluation also effectively shows that extensions to the simple TypeDecl version of TBAA provide significant improvements in accuracy.
+
+
+## Stengths and Weaknesses of the paper
+This TBAA paper took a very concrete approach with examples from Modula-3 syntax, as well as three different performance analysis. One advantage of this is that the paper is fairly self-contained, and doesn't require highly specialized knowledge to understand. Thus the paper is more accessible to researchers from other areas, as well as people who want to implement TBAA but don't have a super strong techincal background. However, as a consequence, the paper doesn't explicitly provide a general theoretical framework that can be easily adopted by future programming languages and concepts.
+
+Limit analysis is a very neat idea, and it should definitely be incorporated into the empirical evaluation of other research projects. Having eight benchmark programs doesn't give the most convincing result, but the fact the programs are all real, practical ones does help.
+
+Finally, what the paper lacks the most is a concrete comparison between the effectiveness of this TBAA against other forms of alias analysis. Although the authors have touched upon how TBAA is usually more efficient, actual numbers and data points would be much more convincing.
+
+
+## TBAA in the Modern Computing Landscape
+
+The simplicity, speed, and accuracy of TBAA has allowed it to stick around in many modern compilers. LLVM has several different alias analyses, including an implementation of TBAA. The compiler frontend must include type annotations that can then be used for alias based optimization within LLVM itself. With a proper understanding of undefined behavior, TBAA can even be implemented for C and C++. Given the continued use of TBAA in real compilers, it is clear that TBAA works well in practice. TBAA can also be used in conjunction with flow sensitive alias analysis to improve accuracy in some cases.
```

Comment:
> With a proper understanding of undefined behavior, TBAA can even be implemented for C and C++.

Not merely *can be* but *is*. I think it is worth noting here that [LLVM](https://llvm.org/doxygen/TypeBasedAliasAnalysis_8cpp_source.html) and [GCC](https://gcc.gnu.org/onlinedocs/gccint/Alias-analysis.html) both implement TBAA.

Reply:
Added these links into the the paragraph